jlm/llvm/opt/alias-analyses/OnlineCycleDetection.hpp

```diff
@@ -0,0 +1,390 @@
+    // Also, if we find any new SCCs while doing this, perform unification right now
+    for (PointerObjectIndex i = 0; i < topologicalOrder.size(); i++)
+    {
+      PointerObjectIndex node = topologicalOrder[i];
+      if (!Set_.IsUnificationRoot(node))
+        continue; // We only care about unification roots
+
+      const auto nextIndex = i + 1;
+
+      // If this root belongs to the same scc as the next root, they should be unified
+      if (nextIndex < topologicalOrder.size()
+          && sccIndex[node] == sccIndex[topologicalOrder[nextIndex]])
+      {
+        // We know that the SCC consists of only roots, since GetSuccessors_ returns only roots
+        JLM_ASSERT(Set_.IsUnificationRoot(topologicalOrder[nextIndex]));
+        // Make the next object the root, to keep the unification going
+        topologicalOrder[nextIndex] = UnifyPointerObjects_(node, topologicalOrder[nextIndex]);
+      }
+      else
+      {
+        // Add this root to the next index in the topological order
+        ObjectToTopoOrder_[node] = TopoOrderToObject_.size();
+        TopoOrderToObject_.push_back(node);
+      }
+    }
+
+    JLM_ASSERT(HasValidTopologicalOrder());
+  }
+
+  /** Call this function after adding a superset edge subset -> superset.
+   * Both must be unification roots.
+   * If subset has a higher topological index than superset, the topological order is re-arranged.
+   * If a cycle is detected, then it will be unified away, and the unification root is returned.
+   * @param subset the tail of the just added superset edge
+   * @param superset the head of the just added superset edge
+   * @return the root of the new unification, if any, otherwise nullopt.
+   */
+  std::optional<PointerObjectIndex>
+  MaintainTopologicalOrder(PointerObjectIndex subset, PointerObjectIndex superset)
+  {
+    JLM_ASSERT(Set_.IsUnificationRoot(subset) && Set_.IsUnificationRoot(superset));
+    // All unification roots should have positions in the topological order
+    JLM_ASSERT(ObjectToTopoOrder_[subset] != -1 && ObjectToTopoOrder_[superset] != -1);
+
+    // If adding the simple edge subset -> superset does not break the invariant, return
+    if (ObjectToTopoOrder_[subset] <= ObjectToTopoOrder_[superset])
+      return std::nullopt;
+
+    // Otherwise, we need to do reordering in this range to fix the invariant
+    const auto lowerTopo = ObjectToTopoOrder_[superset];
+    const auto upperTopo = ObjectToTopoOrder_[subset];
+
+    // Perform DFS to find all roots reachable from superset
+    // We only care about nodes where with its topo index <= upperTopo
+    // These nodes will all need to be moved after subset in the new topological order
+    JLM_ASSERT(DfsStack_.empty());
+    // All nodes that have been pushed to the dfs stack
+    DfsNodesVisited_.Clear();
+    // All nodes that could reach subset and thus are part of a subset -> superset -> subset cycle
+    DfsCycleNodes_.Clear();
```

Comment:
If I understand this correctly, then these methods are only here to ensure that multiple invocations of method can operate correctly. Maybe you can push all these lines in a private method `ClearInternalState()` and invoke it here.